stonesoup/sensormanager/reinforcement.py

```diff
@@ -0,0 +1,294 @@
+                sequence_length=2)
+
+            py_driver.PyDriver(
+                self.env,
+                py_tf_eager_policy.PyTFEagerPolicy(
+                    random_policy, use_tf_function=True),
+                [rb_observer],
+                max_steps=hyper_parameters['initial_collect_steps']).run(train_py_env.reset())
+
+            # Dataset generates trajectories with shape [Bx2x...]
+            dataset = replay_buffer.as_dataset(
+                num_parallel_calls=3,
+                sample_batch_size=hyper_parameters['batch_size'],
+                num_steps=2).prefetch(3)
+
+            iterator = iter(dataset)
+
+            # (Optional) Optimize by wrapping some code in a graph using TF function.
+            self.agent.train = common.function(self.agent.train)
+
+            # Reset the train step.
+            self.agent.train_step_counter.assign(0)
+
+            # Evaluate the agent's policy once before training.
+            avg_return = self.compute_avg_return(self.eval_env, self.agent.policy,
+                                                 hyper_parameters['num_eval_episodes'])
+            returns = [avg_return]
+
+            # Reset the environment.
+            time_step = train_py_env.reset()
+
+            # Create a driver to collect experience.
+            collect_driver = py_driver.PyDriver(
+                self.env,
+                py_tf_eager_policy.PyTFEagerPolicy(
+                    self.agent.collect_policy, use_tf_function=True),
+                [rb_observer],
+                max_steps=hyper_parameters['collect_steps_per_iteration'])
+
+            for _ in range(hyper_parameters['num_iterations']):
+                # Collect a few steps and save to the replay buffer.
+                time_step, _ = collect_driver.run(time_step)
+
+                # Sample a batch of data from the buffer and update the agent's network.
+                experience, unused_info = next(iterator)
+                train_loss = self.agent.train(experience).loss
+
+                step = self.agent.train_step_counter.numpy()
+
+                if step % hyper_parameters['log_interval'] == 0:
+                    print('step = {0}: loss = {1}'.format(step, train_loss))
+
+                if step % hyper_parameters['eval_interval'] == 0:
+                    # Agent Policy Output
+                    avg_return = self.compute_avg_return(self.eval_env, self.agent.policy,
+                                                         hyper_parameters['num_eval_episodes'])
+                    returns.append(avg_return)
+                    print('step = {0}: Average Return = {1}'.format(step, avg_return))
+                    if ('max_train_reward' in hyper_parameters) and\
+                            (avg_return > hyper_parameters['max_train_reward']):
+                        break
+
+            print('\n-----\nTraining complete\n-----')
+
+    def choose_actions(self, tracks, sensor, timestamp, nchoose=1, **kwargs):
```

Comment:
SensorManager base class requires a set of sensors it is tasking so is already aware of them - do the sensor(s) also need to be passed to choose_actions?

Reply:
I've thought about this and initially I didn't pass sensors to this method. The reason I pass sensor is so that when you want to generate an action using the generate_action method in the environment, it's not dependant on the sensor in the class. There's probably a cleverer way of doing this, might need a longer discussion.

Reply:
That makes sense - I think the code here might need adapting for that as I think it will always use the sensor from self.sensors